src/charm.py

```diff
@@ -528,14 +530,61 @@ def _on_start(self, _: StartEvent) -> None:
         # would remain in maintenance status.
         self.unit.status = ActiveStatus()
 
+    def _deduped_relations(self) -> List[ops.Relation]:
+        """Return a relation list with duplicates removed.
+
+        Relations are considered duplicated if they meet the following criteria:
+        1. Two are connected via the legacy ingress and the nginx-route relation endpoint
+        2. Two share the same 4-tuple (service-hostname, service-name, service-model, service-port)
+        3. Both relations have the same remote application.
+
+        In the case of duplicates, the relation connected via legacy ingress is removed.
+
+        Returns:
+            A relation list with duplicates removed.
+        """
+
+        def key(rel: Relation) -> typing.Tuple[str, ...]:
+            """Generate a key from the given relation to detect duplicates.
+
+            Args:
+                rel: the given relation object.
+
+            Returns: The key as a tuple of strings.
+
+            Raises:
+                RuntimeError: if the remote application is unknown.
+            """
+            app = rel.app
+            if not app:
+                raise RuntimeError(f"can't retrieve remote application from relation {rel}")
+            data = rel.data[app]
+            return (
+                data.get("service-hostname"),
+                data.get("service-name"),
+                data.get("service-model"),
+                data.get("service-port"),
+                app.name,
+            )
+
+        nginx_route_relations = self.model.relations["nginx-route"]
+        ingress_relations = self.model.relations["ingress"]
+        nginx_route_relation_keys = set(key(r) for r in nginx_route_relations)
+        dedup_ingress_relations = []
+        for relation in ingress_relations:
```

Comment:
list comprehension?

Reply:
I need to log some message here, so a for loop is more convenient. Thanks!